libclamav/readdb.c

```diff
@@ -581,22 +620,70 @@ cl_error_t readdb_parse_ldb_subsignature(struct cli_matcher *root, const char *v
     return status;
 }
 
+/**
+ * @brief Parse a yara string (subsignature equivalent in yara lingo).
+ *
+ * @param root      The matcher root (engine structure containing loaded signature patterns for matching)
+ * @param virname   Name of signature that this regex subsig came from.
+ * @param hexsig    The string containing the regex
+ * @param subsig_opts Content match pattern options. See ACPATT_* macros in matcher-ac.h.
+ * @param offset    The string offset where the pattern starts
+ * @param target    The clamav target type.
+ * @param lsigid    An array of 2 uint32_t numbers: lsig_id and subsig_id. May be NULL for testing.
+ * @param options   Database options.  See CL_DB_* macros in clamav.h.
+ * @return cl_error_t
+ */
+static cl_error_t readdb_parse_yara_string(struct cli_matcher *root, const char *virname, char *hexsig, uint8_t subsig_opts,
+                                           const char *offset, uint8_t target, const uint32_t *lsigid, unsigned int options)
+{
+    cl_error_t status = CL_EPARSE;
+    cl_error_t ret;
+
+    if (strchr(hexsig, '/')) {
+        /*
+         * Looks like a pcre subsignature.
+         */
+        ret = readdb_load_regex_subsignature(root, virname, hexsig, offset, lsigid, options);
+
+    } else {
+        /*
+         * Looks like an AC/BM content match subsignature.
+         */
+        if (subsig_opts) {
+            ret = cli_sigopts_handler(root, virname, hexsig, subsig_opts, 0, 0, offset, target, lsigid, options);
+        } else {
+            ret = cli_add_content_match_pattern(root, virname, hexsig, 0, 0, 0, offset, target, lsigid, options);
+        }
+    }
+
+    if (CL_SUCCESS != ret) {
+        status = ret;
+        goto done;
```

Comment:
The `goto done` (and `done:`) here could simply be removed (and the trailing `status = CL_SUCCESS` be put in an `else` clause).

But I know we love `done`... so leave it be if you like. But it's nice not to have `goto`'s

For that matter, you could just return directly within the if/else blocks.

Reply:
Yeah I definitely love the goto's because without them, people extend and extend over time with more and more cruft. Error handling gets copy-pasted and then, inevitably, something gets missed and we're left to fix a leak. 
I know it looks kind of stupid in a function this small.  It's really just a future-proofing habit, I guess.